### ISSUE TITLE
Build issue in a nested component fixing branch

### ISSUE THREAD

**Issue by arrowtype** [arrowtype/recursive#427]
Will add more info as I find it.

Current traceback:

```
(venv) 
type-repos/recursive/mastering 
▶ version=1.070 # (replace version number)                            
python build.py --statfiles --version $version
python build.py --static --version $version
🚚 Building files for mastering

🚚 Generating sources
🏗  Copying files
🏗  Opening sources
🏗  Checking family name
🏗  Removing non-exporting glyphs
🏗  Decomposing pair punctuation
🏗  Decomposing problem glyphs
🏗  Clearing guides
🏗  Removing glyphs that aren't in every font
🏗  Removing non-compatible glyphs
🏗  Making kerning compatible
🏗  Sorting glyph order to be common
🏗  Setting production names
🏗  Setting version
🏗  Closing and saving sources
🏗  Checking full design space
🏗  Writing report
✅ Done preparing sources
🏗  Moved features into UFOs

🚚 Making files for static font mastering
🏗  Made folders for static fonts
🏗  Made FontMenuDB
🏗  Made Family features
🏗  Generating instance UFOs
🏗  Getting blueScale
🏗  Setting values, removing overlap, writing files
Traceback (most recent call last):--------------------| 0.0% Complete
  File "build.py", line 78, in <module>
    files = buildFiles(version=version, variable=False)
  File "/Users/stephennixon/type-repos/recursive/mastering/build_files.py", line 107, in buildFiles
    buildInstances(paths["designspace"], paths["cff"], name_map)
  File "/Users/stephennixon/type-repos/recursive/mastering/build_static.py", line 515, in buildInstances
    writeKerning(font, path)
  File "/Users/stephennixon/type-repos/recursive/mastering/build_static.py", line 210, in writeKerning
    w.write(font, feaFile)
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/ufo2ft/featureWriters/baseFeatureWriter.py", line 109, in write
    self.setContext(font, feaFile, compiler=compiler)
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/ufo2ft/featureWriters/kernFeatureWriter.py", line 207, in setContext
    ctx.kerning = self.getKerningData(font, feaFile, self.getOrderedGlyphSet())
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/ufo2ft/featureWriters/baseFeatureWriter.py", line 155, in getOrderedGlyphSet
    skipExportGlyphs=set(font.lib.get("public.skipExportGlyphs", [])),
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/ufo2ft/util.py", line 46, in from_layer
    layer = font.layers.defaultLayer
AttributeError: 'tuple' object has no attribute 'defaultLayer'
🏗  Initial OTF building
   |██████████████████████████████████████████████████| 100.0% Complete
🏗  OTF table fixing
   |██████████████████████████████████████████████████| 100.0% Complete
🏗  OTF autohinting
   |██████████████████████████████████████████████████| 100.0% Complete
🏗  Copying files
🏗  Making TTF sources
Traceback (most recent call last):--------------------| 1.6% Complete
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/fontTools/feaLib/lexer.py", line 258, in __next__
    self.lexers_.append(self.make_lexer_(path))
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/fontTools/feaLib/lexer.py", line 271, in make_lexer_
    fileobj = open(filename, "r", encoding="utf-8")
FileNotFoundError: [Errno 2] No such file or directory: '/Users/stephennixon/type-repos/recursive/mastering/build/static/CFF/RecursiveSansCasualStatic/SemiBoldItalic/./features/common.fea'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "build.py", line 83, in <module>
    build_static(files["cff"], files["ttf"], out)
  File "/Users/stephennixon/type-repos/recursive/mastering/build_static.py", line 768, in build_static
    buildTTFfiles(cff_root, ttf_root)
  File "/Users/stephennixon/type-repos/recursive/mastering/build_static.py", line 167, in buildTTFfiles
    ttf = compileTTF(ufo, useProductionNames=False)
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/ufo2ft/__init__.py", line 233, in compileTTF
    debugFeatureFile=debugFeatureFile,
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/ufo2ft/__init__.py", line 556, in compileFeatures
    otFont = featureCompiler.compile()
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/ufo2ft/featureCompiler.py", line 124, in compile
    self.setupFeatures()
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/ufo2ft/featureCompiler.py", line 219, in setupFeatures
    featureFile = parseLayoutFeatures(self.ufo)
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/ufo2ft/featureCompiler.py", line 46, in parseLayoutFeatures
    doc = parser.parse()
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/fontTools/feaLib/parser.py", line 97, in parse
    statements.append(self.parse_languagesystem_())
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/fontTools/feaLib/parser.py", line 1410, in parse_languagesystem_
    self.expect_symbol_(";")
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/fontTools/feaLib/parser.py", line 1797, in expect_symbol_
    self.advance_lexer_()
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/fontTools/feaLib/parser.py", line 1869, in advance_lexer_
    ) = next(self.lexer_)
  File "/Users/stephennixon/type-repos/recursive/venv/lib/python3.7/site-packages/fontTools/feaLib/lexer.py", line 260, in __next__
    raise IncludedFeaNotFound(fname_token, fname_location) from err
fontTools.feaLib.error.IncludedFeaNotFound: /Users/stephennixon/type-repos/recursive/mastering/build/static/CFF/RecursiveSansCasualStatic/SemiBoldItalic/RecursiveSans-CasualSemiBoldItalic.ufo/features.fea:6:8: The following feature file should be included but cannot be found: ./features/common.fea
```

**Comment by arrowtype** [comment]
Drilling in, it could be that feature code is no longer getting built into UFOs. I need to compare the build process between `main` and this branch.

**Comment by arrowtype** [comment]
One way to visualize a symptom of the problem: with the dependencies updated, sources no longer build to `mastering/build/static/TTF` to prep for the final static build. So, maybe this indicates that the problem could be in AFDKO?

![image](https://user-images.githubusercontent.com/45946693/102414631-e26f0a80-3fc4-11eb-84e8-cf395fda0de9.png)

**Comment by arrowtype** [comment]
I was able to build statics after reverting back to the requirement versions as listed in `requirements.freeze.txt`, and updating just ufo2ft.

However, this currently has an issue with transformed components moving when flattened (https://github.com/googlefonts/ufo2ft/issues/442). So, I will simply decompose a few more glyphs in prep_fonts.py, and I hope that will solve this. Building again now.

**Comment by arrowtype** [comment]
😩 Nooooo I can only build the varible font with ufo2ft==2.19.1, but I can only build the static fonts with ufo2ft==2.15.0.

**Comment by benkiel** [comment]
> One way to visualize a symptom of the problem: with the dependencies updated, sources no longer build to `mastering/build/static/TTF` to prep for the final static build. So, maybe this indicates that the problem could be in AFDKO?

Not in AFDKO, the files for building from the AFDKO aren't being made. It's before that.

**Comment by arrowtype** [comment]
Okay, thank you for clarifying that point, @benkiel!

**Comment by arrowtype** [comment]
I will try to open the font object with ufoLib2, rather than FontParts, to see whether ufo2ft might handle that better.

FontMake does that [like this](https://github.com/googlefonts/fontmake/blob/8705253bc20a6015f81c1c29e3a871fbe88d4b01/Lib/fontmake/font_project.py#L120):

```python
import ufoLib2
    
def open_ufo(self, path):
    try:
        return ufoLib2.Font.open(path, validate=self.validate_ufo)
    except Exception as e:
        raise FontmakeError("Reading UFO source failed", path) from e
```

**Comment by arrowtype** [comment]
Closed/obviated by #428
